Lib/platform.py

```diff
@@ -1441,6 +1442,18 @@ def freedesktop_os_release():
     return _os_release_cache.copy()
 
 
+def invalidate_caches():
```

Comment:
Doesn’t make much sense to me given the context of the original issue #122525 to be clearing all caches in the same function. Entries should only be invalidated when it is correct to invalidate them.

Reply:
No, because some of the caches are related to each others. It's better than to have four functions or two functions that have intricate relations. For instance, if you clear the `uname()` cache, you also need to clear the platform one since the latter queries the former.

Reply:
Well. We could have a separate function invalidating the `os_release` and the `sys_version` cache but I think it's preferrable to invalidate every cached results rather than one by one.

Reply:
For `os_release`, it's still delicate because it relies on `/etc/os-release` which is a simple file. Rather than having possible different values, I prefer invalidating everything simultaneously (though, I may be opened to invalidating `sys_version` separately since it does not depend on external values).

Reply:
Clearing all caches seems like a counter-productive method of solving this. 

> It's better than to have four functions or two functions that have intricate relations.

I don’t believe that the current implementation is better than this.

Reply:
Could you then develop your argument and suggest alternatives? We want to have cached calls because most of the time the results will not change during the execution. In addition, we definitely don't want to make them non-cached since changing the platform's node is not a frequent task. It remains to decide whether we want to invalidate caches (which are implementation details by the way) in one go (which is what I think it should do) or require the user to remember different functions (which we will need to teach as well).

I agree on splitting the implementation for `os_release` and `sys_version`, but platforms and unames caches MUST be cleared together.

Reply:
I think it's a great idea to give an option to invalidate caches. While yes, it doesn't solve the original issue directly, #122525 had `socket.gethostname` as a non-caching alternative. 

> Entries should only be invalidated when it is correct to invalidate them.

Not disagreeing in the sense that it's good API design, but by adding cache invalidation to `uname` we'll be slowing it down for a very niche feature -- typically, `platform` is used for things that don't really care about caches e.g. printing info about the current system for debugging. `invalidate_caches` defers the performance tradeoff to the user, if they really need the most recent information.

Anyways, if we *really* wanted to add cache invalidation internally, it should only be in `node` (and possibly `version` and `release`), not `uname`. I don't think anyone is switching out their processor in the middle of execution -- users who are using `uname` (directly or indirectly, through `platform.processor()`) for finding the processor would be affected by adding cache invalidation for other values there.

Reply:
Here we are clearing all 4 entries of the uname cache which I believe is different from what is happening in those other functions. Although I am not sure how expensive it is to repopulate said uname cache so maybe the current idea for this PR is fine.

Reply:
> Here we are clearing all 4 entries of the uname cache which I believe is different from what is happening in those other functions

Some other clean-up functions (such as `urllib.request.urlcleanup`) clear more than just one cache (and the caches are not related to each others). More precisely, we are not clearing 4 entries of the uname cache. What we are clearing are: `uname()`, `platform()` (which depends on `uname()`, so it should be cleared if `uname()` is), `freedesktop_os_release()`, which relies on `/etc/os-release` file which can be modified externally and the cache about Python version `_sys_version()`.

Perhaps I should have left out the `freedesktop_os_release` and the `sys_version` caches because:

- AFAICT, the only way to have outdated `platform.freedesktop_os_release()` is if you dist-upgrade your current system while having a Python interpreter with a cached call to `platform.freedesktop_os_release()`. If you don't reboot (which I think is possible with some distros that do live patching), and if the upgrade actually updated `/etc/os-release`,  then `platform.freedesktop_os_release()` would have an oudated distribution version (probably). This is really a corner case which should probably never happen in production, so I could definitely assume that the cache wouldn't need to be cleared up.
- the sys version because the information cannot be changed by an external process (I think). I can remove that cache clearance because this one might be costly to reconstruct (due to regexes).
 
@vstinner Do you think I should have left out clearing up the cache for `/etc/os-release` and the cache for the sys version since the results are rarely (or never) changed be external processes during the interpreter lifetime? By doing so, the docs should specify *which* caches are to be cleared (which is maybe something that the users don't want to bother about).

----

Actually, there is a more general way to invalidate the caches of modules, and this is to reload the module itself. Not all modules with cached objects expose cache clearing mechanism (for instance, `urllib.request.urlcleanup` does not clear the cached host IP address) but most of the time, the important caches can be purged by one function (and not by a bunch of functions).

> Although I am not sure how expensive it is to repopulate said uname cache so maybe the current idea for this PR is fine

Here, performance don't really matter because clearing the cache is not something you would do a lot of time. You would do cache clearing if you really want to be sure that the information is up-to-date if another process might have change it. If you want to do it many times, then this means that you probably didn't want a cache in the first place.

Reply:
It makes sense to invalidate `/etc/os-release` cache since this file can be replaced/updated by a system upgrade while a Python process is running.

Reply:
I'll just keep all caches since someone could manually change the values of `sys.version`, `sys.platform` and `sys._git` or `sys._mercurial`.

Reply:
> someone could manually change the values of sys.version, sys.platform and sys._git or sys._mercurial

You're not supposed to do that. If you do, you're on your own :-)

Reply:
No indeed.. anyway, I'll just keep the cache clear for the sys-related objects since otherwise the docs need to be improved saying that some functions will not have their cache cleared and I think it will cause more confusion.